drivers: ssd1306 Add SPI support.

Marco-Peter opened the pull request:
This change allows to use SSD1306 based displays to be used on the SPI bus
as well.
Additionally, the interleaving control frames are removed to get
the same data format on both buses. This leads to an increased
transfer speed on I2C as well.
Adding SPI shield.
Tested on SSD1306 and SSD1309 based displays using I2C.
Tested on SSD1309 based display using SPI.

Signed-off-by: Marco Peter <marco@peter-net.ch>

jfischer-phytec-iot commented on drivers/display/ssd1306.c:
```diff
@@ -48,144 +49,130 @@ LOG_MODULE_REGISTER(ssd1306, CONFIG_DISPLAY_LOG_LEVEL);
-				 reg, val);
-}
+	struct ssd1306_data *driver = dev->data;
+#if DT_INST_ON_BUS(0, i2c)
```
I would rather see 
```c
#if DT_INST_ON_BUS(0, i2c)                                                                    
static inline int ssd1306_write_bus(const struct device *dev,                                 
                                    uint8_t *buf, size_t len, bool command)                   
{                                                                                             
        struct ssd1306_data *driver = dev->data;                                              
                                                                                              
        return i2c_burst_write(driver->bus, DT_INST_REG_ADDR(0),                              
                               command ? SSD1306_CONTROL_ALL_BYTES_CMD :                      
                                         SSD1306_CONTROL_ALL_BYTES_DATA,                      
                               buf, len);                                                     
} 
                                                                                            
#elif DT_INST_ON_BUS(0, spi)    
                                                              
static inline int ssd1306_write_bus(const struct device *dev,                                 
                                    uint8_t *buf, size_t len, bool command)                   
{                                                                                             
        struct ssd1306_data *driver = dev->data;                                              
        int errno;                                                                            
                                                                                              
        gpio_pin_set(driver->data_cmd, DT_INST_GPIO_PIN(0, data_cmd_gpios),                   
                     command ? 0 : 1);                                                        
        struct spi_buf tx_buf = {                                                             
                .buf = buf,                                                                   
                .len = len                                                                    
        };                                                                                    
                                                                                              
        struct spi_buf_set tx_bufs = {                                                        
                .buffers = &tx_buf,                                                           
                .count = 1                                                                    
        };                                                                                    
                                                                                              
        errno = spi_write(driver->bus, &driver->spi_config, &tx_bufs);                        
        gpio_pin_set(driver->data_cmd,                                                        
                     DT_INST_GPIO_PIN(0, data_cmd_gpios), 0);                                 
                                                                                              
        return errno;                                                                         
}                                                                                             
#endif
```

Marco-Peter replied:
done

jfischer-phytec-iot commented on drivers/display/ssd1306.c:
```diff
@@ -48,144 +49,130 @@ LOG_MODULE_REGISTER(ssd1306, CONFIG_DISPLAY_LOG_LEVEL);
 {
-	struct ssd1306_data *driver = dev->data;
 	uint8_t cmd_buf[] = {
-		SSD1306_CONTROL_BYTE_CMD,
```
Please split your PR in few commits and do this optimization (and related [changes](https://github.com/zephyrproject-rtos/zephyr/pull/28322/commits/a53c308e1458fadf31e8a2561a6ddaf3e5143bf6#diff-5fd46386a4f473a41f96f54a41f497e3R11)) as a first step.

Marco-Peter replied:
For now I just splitted the PR Into 2 commits, within the current PR:
- optimize protocol
- add SPI support
If requested, I can also add more splits or split the PR itself.

Marco-Peter committed:
drivers: ssd1306: Optimize display commands

This change removes the interleaving control
frames.
Additionally all I2C accesses are centralized in
one single function.

Signed-off-by: Marco Peter <marco@peter-net.ch>

Marco-Peter committed:
drivers: ssd1306: Add SPI support

This change allows to use SSD1306 based displays to be used on the
SPI bus as well.
Adding SPI shield.
Tested on SSD1306 and SSD1309 based displays using I2C.
Tested on SSD1309 based display using SPI.

Signed-off-by: Marco Peter <marco@peter-net.ch>

jfischer-phytec-iot commented on drivers/display/ssd1306.c:
```diff
@@ -371,6 +406,34 @@ static int ssd1306_init(const struct device *dev)
 #endif
 
+#if DT_INST_ON_BUS(0, spi)
+	driver->spi_config.frequency = 10000000;
```
driver->spi_config.frequency = DT_INST_PROP(0, spi_max_frequency);

Marco-Peter replied:
Is corrected now.

jfischer-phytec-iot commented on drivers/display/ssd1306.c:
```diff
@@ -48,22 +49,56 @@ LOG_MODULE_REGISTER(ssd1306, CONFIG_DISPLAY_LOG_LEVEL);
+	};
+
+	errno = spi_write(driver->bus, &driver->spi_config, &tx_bufs);
+	gpio_pin_set(driver->data_cmd,
```
Is it necessary to reset DC gpio?

Marco-Peter replied:
No, since the pin usually is not pulled in any direction, it is perfectly safe to let it be.
I removed the command.

vanwinkeljan commented on drivers/display/Kconfig.ssd1306:
```diff
@@ -5,7 +5,7 @@
 menuconfig SSD1306
 	bool "SSD1306 display driver"
-	depends on I2C
+	depends on (I2C && HAS_DTS_I2C) || SPI
```
Remove HAS_DTS_I2C it is no longer needed, https://github.com/zephyrproject-rtos/zephyr/pull/28517

Marco-Peter replied:
done

erwango commented on boards/shields/ssd1306/Kconfig.defconfig:
```diff
@@ -1,13 +1,24 @@
 
-if SHIELD_SSD1306_128X64 || SHIELD_SSD1306_128X32 || SHIELD_SH1106_128X64
-
 if DISPLAY
 
+if SHIELD_SSD1306_128X64 || SHIELD_SSD1306_128X64_SPI || SHIELD_SSD1306_128X32 || SHIELD_SH1106_128X64
```
Can you explain why you're inverting the lines here ?

Marco-Peter replied:
Basic idea was to have a hierarchical ordering from the most generic (DISPLAY) down to the actual types and their configurations (SPI or I2C).

erwango replied:
> Basic idea was to have a hierarchical ordering from the most generic (DISPLAY) down to the actual types and their configurations (SPI or I2C).

Ok, I agree on the general idea. Though in this concept, I think the highest prio would be: hardware is present (so one of `SHIELD_SSD1306_FOO` is true). So nothing in this file should be treated if this condition doesn't apply.
This is also quite easy to explain and apply. 
Last, this change breaks consistency with other shields, so it would have to be done in a dedicated PR, and take all shields into account.

erwango commented on boards/shields/ssd1306/Kconfig.defconfig:
```diff
@@ -1,13 +1,24 @@
 
-if SHIELD_SSD1306_128X64 || SHIELD_SSD1306_128X32 || SHIELD_SH1106_128X64
-
 if DISPLAY
 
+if SHIELD_SSD1306_128X64 || SHIELD_SSD1306_128X64_SPI || SHIELD_SSD1306_128X32 || SHIELD_SH1106_128X64
```
Please keep "hardware present" as the highest prio condition.
